File: javascript/data-structures/p164_maxPQ.js
/*
    CLRS Section 6.5, p. 164
*/

const { MaxHeap } = require('../data-structures/p157_maxHeap');

class MaxPQ extends MaxHeap {
    constructor(A, heapSize) {
        super(A, heapSize);
    }

    heapMaximum = () => this.A[0];

    heapExtractMax = () => {
        if (this.heapSize < 1) {
            throw new Error('heap underflow');
        }
        const max = this.heapMaximum();
        this.A[0] = this.A.pop();
        this.heapSize--;
        this.maxHeapify(0);
        return max;
    }

    heapIncreaseKey = (i, key) => {
        if (key < this.A[i]) {
            throw new Error('new key is smaller than current key');
        }
        this.A[i] = key;
        while (i > 0 && (this.A[this.parent(i)] < this.A[i])) {
            [this.A[i], this.A[this.parent(i)]] = [this.A[this.parent(i)], this.A[i]];
            i = this.parent(i);
        }
    }

    maxHeapInsert = (key) => {
        this.heapSize++;
        this.A.push(Number.NEGATIVE_INFINITY);
        this.heapIncreaseKey(this.heapSize - 1, key);
    }

    // N.B. cf. Exercise 6.5-8, p. 166
    heapDelete = (i) => { // TO-DO: not correct for ELSE case
        // if (this.A[i] > this.A[this.heapSize - 1]) {
        //     this.A[i] = this.A.pop();
        //     this.maxHeapify(i);
        // } else {
        //     if (i < this.heapSize - 1) {
        //         const key = this.A.pop();
        //         this.heapIncreaseKey(i, key);
        //     } else {
        //         this.heapIncreaseKey(i, this.A[this.heapSize - 1]);
        //     }
        // }
        // this.heapSize--;
    } 
}

module.exports = {
    MaxPQ
}

File: python/data_structures/p164_max_pq.py
'''
    CLRS Section 6.5, p. 164
'''

from .p157_max_heap import MaxHeap
import math

class MaxPQ(MaxHeap):
    def __init__(self, A, heap_size):
        super().__init__(A, heap_size)
    
    def heap_maximum(self):
        return self.A[0]
    
    def heap_extract_max(self):
        if self.heap_size < 1:
            raise Exception('heap underflow')
        max = self.heap_maximum()
        self.A[0] = self.A.pop()
        self.heap_size -= 1
        self.max_heapify(0)
        return max
    
    def heap_increase_key(self, i, key):
        if key < self.A[i]:
            raise Exception('new key is smaller than current key')
        self.A[i] = key
        while i > 0 and self.A[self.parent(i)] < self.A[i]:
            self.A[i], self.A[self.parent(i)] = self.A[self.parent(i)], self.A[i]
            i = self.parent(i)
    
    def max_heap_insert(self, key):
        self.heap_size += 1
        self.A.append(-math.inf)
        self.heap_increase_key(self.heap_size - 1, key)
    
    # N.B. cf. Exercise 6.5-8, p. 166
    def heap_delete(self, i):
        pass # TO-DO
